#### Ad Soyad:

In [ ]:
import pandas as pd

### 0. MarketSales.xlsx dosyasını linkten indirin: https://www.kaggle.com/omercolakoglu/turkish-market-sales-dataset-with-9000items

In [2]:
df = pd.read_excel('MarketSales.xlsx')
df['DATE_'] = pd.to_datetime(df['DATE_'])

In [3]:
df.head()

ID  ITEMCODE                         ITEMNAME  FICHENO      DATE_  \
0  11738.0    5863.0  SPRITE 1 LT LIMON AROMALI GAZOZ    18456 2017-01-07   
1  10537.0       8.0                        TOZ SEKER    18105 2017-01-06   
2  11335.0    5979.0             FALIM SAKIZ 5LI NANE    18350 2017-01-03   
3  11336.0    5979.0             FALIM SAKIZ 5LI NANE    18350 2017-01-03   
4  10115.0   12808.0            FALIM SAKIZ 5LI CILEK    18005 2017-01-05   

   AMOUNT  PRICE  LINENETTOTAL  LINENET  BRANCHNR  ... CLIENTCODE  \
0     1.0   2.00          2.00     1.85      52.0  ...     467369   
1     5.0   2.65         13.25    12.27       8.0  ...     131464   
2     1.0   0.40          0.40     0.37      40.0  ...     656969   
3     1.0   0.40          0.40     0.37      40.0  ...     656969   
4     1.0   0.40          0.40     0.37      41.0  ...     447336   

       CLIENTNAME BRANDCODE   BRAND  CATEGORY_NAME1   CATEGORY_NAME2  \
0  Sercan KIZILOK       156  SPRİTE          İÇECEK     GAZLI İÇECEK   
1    İsmet ŞINGIR       NaN     NaN          İÇECEK        ÇAY KAHVE   
2     Yağız KUBAL       300   FALIM            GIDA  SAKIZ SEKERLEME   
3     Yağız KUBAL       300   FALIM            GIDA  SAKIZ SEKERLEME   
4    Habibe AYSAN       300   FALIM            GIDA  SAKIZ SEKERLEME   

       CATEGORY_NAME3           STARTDATE             ENDDATE GENDER  
0               GAZOZ 2017-01-08 16:16:11 2017-01-08 16:17:13      E  
1  SEKER TATLANDIRICI 2017-01-07 11:04:34 2017-01-07 11:05:37      E  
2               SAKIZ 2017-01-04 14:00:03 2017-01-04 14:01:01      E  
3               SAKIZ 2017-01-04 14:00:03 2017-01-04 14:01:01      E  
4               SAKIZ 2017-01-06 14:00:30 2017-01-06 14:01:03      K  

[5 rows x 26 columns]

### 1. Veri setini inceleyin ve kolonları anlamaya çalışın. Bu adımda bir çıktı/kod göstermek zorunda değilsiniz. Sadece veriyi anlamak için...

In [57]:
df.columns

Index(['ID', 'ITEMCODE', 'ITEMNAME', 'FICHENO', 'DATE_', 'AMOUNT', 'PRICE',
       'LINENETTOTAL', 'LINENET', 'BRANCHNR', 'BRANCH', 'SALESMAN', 'CITY',
       'REGION', 'LATITUDE', 'LONGITUDE', 'CLIENTCODE', 'CLIENTNAME',
       'BRANDCODE', 'BRAND', 'CATEGORY_NAME1', 'CATEGORY_NAME2',
       'CATEGORY_NAME3', 'STARTDATE', 'ENDDATE', 'GENDER'],
      dtype='object')

Mesela 7 numaralı şubede günlük toplam tereyağ satış miktarını şu şekilde görüyoruz:

In [35]:
df[(df['BRANCHNR']==7) & (df['CATEGORY_NAME3']=='TEREYAĞ')].groupby('DATE_').sum()['AMOUNT']

DATE_
2017-01-04    1.0
2017-01-05    1.0
2017-01-06    1.0
2017-01-07    5.0
2017-01-08    2.0
             ... 
2017-03-27    2.0
2017-03-28    2.0
2017-03-29    1.0
2017-03-30    3.0
2017-03-31    2.0
Name: AMOUNT, Length: 74, dtype: float64

656969 numaralı müşterinin alışveriş başına harcadığı miktarları şu şekilde görebiliriz:

In [49]:
df[(df['CLIENTCODE']=='461524')].groupby('DATE_').sum()['PRICE']

DATE_
2017-01-08    101.67
2017-01-19      4.75
2017-02-19      2.30
2017-02-25     97.22
2017-03-08    363.40
2017-03-31      7.95
Name: PRICE, dtype: float64

Şubeleri ve her bir şubenin kaç dataya sahip olduğunu görmek için:

In [58]:
df['BRANCH'].value_counts()

İstanbul Subesi     113622
Ankara Subesi        40695
İzmir Subesi         32328
Bursa Subesi         21215
Antalya Subesi       17371
                     ...  
Kilis Subesi          1195
Gümüşhane Subesi      1187
Ardahan Subesi         734
Tunceli Subesi         563
Bayburt Subesi         528
Name: BRANCH, Length: 81, dtype: int64

### 2.a. En çok para harcayan müşteriyi (CLIENTCODE ve CLIENTNAME) bulunuz. Bu müşterinin en çok harcama yaptığı kategorileri (CATEGORY_NAME1, CATEGORY_NAME2, CATEGORY_NAME3 için ayrı ayrı) gösteriniz. Pie chart veya bar chart ile görselleştirme yapınız.

### 2.b. Toplam cirosu en fazla olan mağazayı bulunuz. Bu mağazada en çok satılan kategorileri (CATEGORY_NAME1, CATEGORY_NAME2, CATEGORY_NAME3 için ayrı ayrı) gösteriniz.  Pie chart veya bar chart ile görselleştirme yapınız.

### 2.c. Satışı en fazla olan 3 ürünü (ITEMNAME) bulunuz. Bu ürünleri en çok satın alan (AMOUNT) 3 müşteriyi (CLIENTCODE ve CLIENTNAME) gösteriniz. Bu ürünün en fazla satıldığı 3 şubeyi (BRANCHNR ve BRANCH) bulunuz.  Pie chart veya bar chart ile görselleştirme yapınız.

### [Toplamda 9 müşteri ve 9 mağaza bulacaksınız.]

### 2.d. Satışı en fazla olan 3 markayı (BRAND) bulunuz. Her bir marka için ayrı ayrı olmak üzere, bu markaları en çok tüketen (AMOUNT) 3 müşteriyi (CLIENTCODE ve CLIENTNAME) gösteriniz. Bu markaların en fazla satıldığı 3 şubeyi (BRANCHNR ve BRANCH) bulunuz. Pie chart veya bar chart ile görselleştirme yapınız.

### [Toplamda 9 müşteri ve 9 mağaza bulacaksınız.]

### 2.e. En çok ürünün satıldığı 3 tarihi bulunuz. Bu tarihler için ayrı ayrı en çok satılan 3 ürünü gösteriniz. Pie chart veya bar chart ile görselleştirme yapınız.
### [Toplamda 9 ürün göstereceksiniz.]

### 3.a. En fazla işlem yapan şubeyi bulunuz. (İpucu: FICHENO üzerinden gruplayarak bulunabilir.)

In [55]:
for bnr in df.BRANCHNR.unique():
    print(bnr,len(df[df.BRANCHNR==bnr].groupby('FICHENO').sum()))

52.0 3271
8.0 4050
40.0 26360
41.0 7507
29.0 972
55.0 1411
56.0 2464
59.0 1577
60.0 702
1.0 3884
66.0 1696
81.0 3347
12.0 2106
48.0 503
33.0 3494
64.0 922
58.0 2974
46.0 611
14.0 1011
73.0 1737
71.0 3440
62.0 576
7.0 9440
32.0 1435
26.0 2946
38.0 334
39.0 824
21.0 5144
42.0 1915
25.0 1787
57.0 1374
3.0 1248
31.0 1343
70.0 1019
37.0 2736
63.0 1253
72.0 881
77.0 652
53.0 3763
24.0 982
67.0 2283
28.0 671
22.0 870
47.0 2358
11.0 1887
19.0 548
34.0 755
74.0 1092
2.0 1013
50.0 440
80.0 705
75.0 1301
78.0 1967
4.0 923
54.0 986
45.0 475
35.0 287
69.0 360
61.0 505
43.0 401
15.0 123
6.0 633
68.0 526
20.0 470
5.0 729
65.0 576
10.0 301
49.0 636
36.0 504
27.0 601
51.0 272
16.0 373
30.0 417
23.0 345
13.0 329
79.0 405
18.0 617
44.0 434
17.0 483
nan 0
9.0 175
76.0 144


### 3.b. 3.a'da bulduğunuz şubeyi tarihe göre gruplandırın ve CATEGORY_NAME3'ten bir ürün grubu seçin. Burada tamamen özgürsünüz. Aşağıda rastgele bir şube ve ürün grubuyla bir örnek gösterilmiştir.

### Bu ürünün satış miktarı (AMOUNT) üzerinden bir LSTM tahmin yapısı kurun. Verinin %80'ini train, %20'sini test datası olarak kullanın. MSE değerini gösterin ve train-test-prediction datalarını içeren grafiği çizin.

### LSTM için farklı lookback (history) değerleri ve neural network size'ı seçebilirsiniz. Bulduğunuz en optimal değeri gösteriniz.

### Özellikle test ve prediction datasını üst üste çizdirin.

In [56]:
df[(df['BRANCHNR']==7) & (df['CATEGORY_NAME3']=='TEREYAĞ')].groupby('DATE_').sum()['AMOUNT']

DATE_
2017-01-04    1.0
2017-01-05    1.0
2017-01-06    1.0
2017-01-07    5.0
2017-01-08    2.0
             ... 
2017-03-27    2.0
2017-03-28    2.0
2017-03-29    1.0
2017-03-30    3.0
2017-03-31    2.0
Name: AMOUNT, Length: 74, dtype: float64

### 3.c. 3.b.'deki örnekte seçtiğiniz şube ve ürün grubu için LSTM yerine FBProphet ile tahminleme yapın. Prophet'in grafiklerini çizdirin.

### Özellikle test ve prediction datasını üst üste çizdirin.

### 4. Hangi model daha iyi sonuç vermiştir? Neden? Sonuçlarınızı tartışınız.